src/java/org/apache/cassandra/db/virtual/ExceptionsTable.java

```diff
@@ -0,0 +1,323 @@
+     * @param occurrenceTime    time when given exception ocurred
+     */
+    private void add(String exceptionClass,
+                     String exceptionLocation,
+                     String message,
+                     List<String> stackTrace,
+                     long occurrenceTime)
+    {
+        synchronized (buffer)
+        {
+            Map<String, ExceptionRow> exceptionRowWithLocation = buffer.computeIfAbsent(exceptionClass, (classToAdd) -> new LinkedHashMap<>());
+            ExceptionRow exceptionRow = exceptionRowWithLocation.get(exceptionLocation);
+            if (exceptionRow == null)
+            {
+                // exception class and location can be null for value as we have it as part of keys already
+                exceptionRow = new ExceptionRow(null, null, 1, message, stackTrace, occurrenceTime);
+                exceptionRowWithLocation.put(exceptionLocation, exceptionRow);
+                // not important, can be null
+                // we need to do this, because if we add into a map which is
+                // a value of some buffer key, we might exceed the number
+                // of overall entries in the buffer
+                buffer.removeEldestEntry(null);
+            }
+            else
+            {
+                exceptionRow.count += 1;
+                exceptionRow.message = message;
+                exceptionRow.stackTrace = stackTrace;
+                exceptionRow.occurrence = new Date(occurrenceTime);
+            }
+        }
+    }
+
+    // public to make warning go away about public preInitialisationBuffer exposing List<ExceptionRow>
+    // where ExceptionRow would have lower visibility than that.
+    public static final class ExceptionRow
+    {
+        final String exceptionClass;
+        final String exceptionLocation;
+        int count;
+        String message;
+        List<String> stackTrace;
+        Date occurrence;
+
+        /**
+         * @param exceptionClass    exception class of uncaught exception
+         * @param exceptionLocation location where that exception was thrown
+         * @param message           message of given exception
+         * @param stackTrace        whole stacktrace of given exception
+         * @param occurrenceTime    time when given exception ocurred, in milliseconds from epoch
+         */
+        ExceptionRow(String exceptionClass,
+                     String exceptionLocation,
+                     int count,
+                     String message,
+                     List<String> stackTrace,
+                     long occurrenceTime)
+        {
+            this.exceptionClass = exceptionClass;
+            this.exceptionLocation = exceptionLocation;
+            this.count = count;
+            this.stackTrace = stackTrace;
+            this.message = message;
+            this.occurrence = new Date(occurrenceTime);
+        }
+
+        public static List<String> extractStacktrace(StackTraceElement[] stackTraceArray)
```

Comment:
Nit: probably the method is better to place in the upper class, it is not used in this one and declarations to use it are longer now..